### PR TITLE
storage: harden artifical quiesce heartbeat

### PR DESCRIPTION
**Pull request by spencerkimball**
We previously had the assumption when sending quiesce messages
that the Commit field could always be set to the Raft group's
`status.Commit`. With upcoming changes to quiesce ranges even
with replicas that are behind but non-live, this value could be
set incorrectly and still received by a supposedly dead replica.

This change mirrors the logic in the raft implementation for
setting the `raftpb.Message.Commit` field.

Release note: None

### CODE REVIEW

**Comment by nvanbenschoten** [comment]
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/storage/replica.go, line 4168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26908#-LFZSLkd5nc9lv-Cl6px:-LFZSLkeB1LyzUxBcAU2:b6vfp6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/064bd505926b98b3c462c94bbc869623771b285c/pkg/storage/replica.go#L4168)):*
> ```Go
> 			Type:   raftpb.MsgHeartbeat,
> 			Term:   status.Term,
> 			Commit: min(prog.Match, status.Commit),
> ```

Leave a comment above this line pointing into Raft at `raft.sendHeartbeat`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26908)*

File: pkg/storage/replica.go [hunk 1]
```diff
@@ -4157,7 +4165,7 @@ func (r *Replica) quiesceAndNotifyLocked(ctx context.Context, status *raft.Statu
 			To:     id,
 			Type:   raftpb.MsgHeartbeat,
 			Term:   status.Term,
-			Commit: status.Commit,
+			Commit: min(prog.Match, status.Commit),
```
Comment (tschottdorf) [inline pkg/storage/replica.go]: Please add a comment such as this one:

// In common operation, we only quiesce when all followers are up-to-date, and
// so the status.Commit field is safe to send. However, when we quiesce in the
// presence of dead nodes, a follower which is behind and counted as dead may
// not have the log entry referenced by `status.Commit` and would explode if it
// were told to commit up to that point. So reduce the commit index to an index
// we know the follower has acknowledged.

**Review by tbg — APPROVED** [review]
@bdarnell I think when I ran into this during my quiesce experiments, you mentioned that it was odd (?) to send a Commit update here in the first place. Is that correct? ISTM that if we didn't do that we could sometimes end up quiescing a follower that has the latest log index but not letting it know it's committed.

**Comment by nvanbenschoten** [comment]
> If we quiesce a replica that doesn't have the up-to-date commit information, it has no way of catching up later unless the replica unquiesces.

This makes sense to me, but I want to double check that I understand the desired behavior here. We don't want to send the heartbeat to the replica with a stale commit index because we never want that replica to quiesce, right? We either want it to be dead or to unquiesce the range until it is caught up?

**Comment by bdarnell** [comment]
Right. If it's dead, it doesn't matter whether we send the heartbeat or not. If it's alive, we want it to remain unquiesced until it catches up.

**Comment by nvanbenschoten** [comment]
> If it's alive, we want it to remain unquiesced until it catches up.

And what actions will it take if it's unquiesced and alive while the rest of the range is quiesced to try to catch up? Will it campaign (non-disruptively and unsuccesfully because of pre-vote) and wake up the other replicas, allowing the leader to continue to catch it up with MsgApps?

**Comment by bdarnell** [comment]
Yes, exactly.

File: pkg/storage/replica.go (hunk 1 above)
Comment (spencerkimball) [reply pkg/storage/replica.go]: I've incorporated this into the new mechanism suggested by @bdarnell.

**Review by spencerkimball — COMMENTED** [review]
OK, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26908#-:-LHAIlLpEqjja7qamQkc:blsgobu)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 4168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26908#-LFZSLkd5nc9lv-Cl6px:-LHAIneK3mwZdjiOyWVk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/064bd505926b98b3c462c94bbc869623771b285c/pkg/storage/replica.go#L4168)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Leave a comment above this line pointing into Raft at `raft.sendHeartbeat`.
</blockquote></details>

Done.

**Review by bdarnell — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26908#-:-LHEb68I68yH7PAqvbd-:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by nvanbenschoten** [comment]
@bdarnell the original code here would have avoided the issue we see in https://github.com/cockroachdb/cockroach/issues/30064#issuecomment-427757319. It was changed because of https://github.com/cockroachdb/cockroach/pull/26908#issuecomment-400166031. We can't revert to the original code because the hazard mentioned in that comment is very real, but I think the correct fix is to send the heartbeat to the straggler replica (which we think is dead) without specifying that the Replica should quiesce (`Quiesce: false`). Doing so means that the straggler Replica will still get a heartbeat if it needs one to join the Raft group (see https://github.com/cockroachdb/cockroach/issues/30064#issuecomment-427764605) and that it will wake the Range up if it happens to come back to life so that it can catch up. Thoughts?

**Review by bdarnell — COMMENTED** [review]
SGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26908#-:-LOK0ci514mgmn17uomk:bb73djm)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Pull request merged**